Add option for node CIDR allocation to kubeadm

taimir opened the pull request:
@errordeveloper Hi, I've been following the sig-cluster-lifecycle for a while, and I am interested in what you guys are doing :) I like it a lot.

I've been testing out `kubeadm` for the last couple of days and I managed to provision an operational cluster on 2 Ubuntu Xenial nodes. However, I had to introduce the following changes in the code:
- For some reason the Services CIDR was passed as Cluster CIDR to the controller-manager. Cluster CIDR is there for the pod network, however, so I don't think that was OK. A lot of the current provisioning projects (e.g. bootkube, kube-deploy/docker-multinode) use flannel as an overlay, and without my changes  flannel won't work
- There was a small bug in the `manual bootstrap`, when the bearer token is referenced in the code.

If you need help on any open topics just ping me, I'd be glad to contribute :) 

@kenan435 @cheld

errordeveloper commented on pkg/kubeadm/cmd/manual.go:
```diff
@@ -216,7 +222,7 @@ func NewCmdManualBootstrapJoinNode(out io.Writer, s *kubeadmapi.KubeadmConfig) *
 		`Comma separated list of API server URLs. Typically this might be just
 		https://<address-of-master>:8080/`)
-	cmd.PersistentFlags().StringVarP(&s.ManualFlags.BearerToken, "token", "", "",
+	cmd.PersistentFlags().StringVarP(&s.Secrets.BearerToken, "token", "", "",
```
Well spotted, thanks!

errordeveloper commented on pkg/kubeadm/master/manifests.go:
```diff
@@ -216,7 +216,8 @@ func getComponentCommand(component string, s *kubeadmapi.KubeadmConfig) (command
 			"--insecure-experimental-approve-all-kubelet-csrs-for-group=system:kubelet-bootstrap",
-			"--cluster-cidr=" + s.InitFlags.Services.CIDR.String(),
+			"--allocate-node-cidrs=true",
+			"--cluster-cidr=" + s.InitFlags.PodNetwork.CIDR.String(),
```
I think this should be optional, it really depends on the choice of network and user has to mach network configuration to reflect this.

errordeveloper replied:
Also, I have to say that not separating these two flags seems natural, but it'd be good to add a comment on the reasoning why we grouped these two (stating what are the use-cases when you want to set it, but don't want to allocate it).

taimir replied:
Agreed, I'll see to this ASAIC.

errordeveloper reviewed (changes requested):
Thanks, I think this is very good. The only thing that I think would be better is to not have a default.

errordeveloper commented:
It is true that flannel needs this, but Weave needs the opposite, and Calico probably also, as they have their own IPAM layer.

Commit:
Add node CIDR allocation as an option to kubeadm.

This is useful for users who are used to deploying with a flannel
overlay network.

taimir commented:
I've updated this PR, PTAL. Now the cluster CIDR should be optional and the user can specify it if needed.

About the grouping of `--alocate-node-cidrs` and `--cluster-cidrs`: AFAICS, the `--allocate-node-cidrs` produces an [error](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/nodecontroller.go#L205) in the NodeController if `--cluster-cidr` is not set. On the other hand `--cluster-cidr` in the NodeController is currently only used in the context of `NewCIDRRangeAllocator()`, called in this [if block](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/nodecontroller.go#L326), called [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/nodecontroller.go#L344). So it seems that those two go hand in hand right now.

The only other place where the `--cluster-cidr` flag is used is in the RouteController, to create routes on the cloud provider plane pointing to the Nodes that the NodeController has allocated IP's on the cloud provider for.

How would you want this documented? I hope I am not missing anything at this point.

errordeveloper commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -65,22 +65,27 @@ func NewCmdInit(out io.Writer, s *kubeadmapi.KubeadmConfig) *cobra.Command {
+	)
+	cmd.PersistentFlags().IPNetVar(
+		&s.InitFlags.PodNetwork.CIDR, "cluster-cidr", net.IPNet{},
+		`(optional) Use alternative range of IP addresses for the pod overlay network VIPs, e.g. 10.3.0.0/16.`+
```
Pod IPs aren't virtual, so using "VIPs" here is misleading...

errordeveloper replied:
Also, strictly speaking it wouldn't always mean that these IPs are used for overlay networks as such, e.g. in the where routes are setup in the cloud cloud provider.

taimir replied:
Sorry about the VIPs, it's a copy-paste residual. 

> e.g. in the where routes are setup in the cloud cloud provider.

I want to understand this correctly - by this you mean that the RouteController can still use the `clusterCIDR` to establish routes to existing cloud provider instances (separate from the cluster) so that the cluster can talk to them? Is that the use case? Or am I misinterpreting this part ... 

I just have not had this flag used for anything else but overlays in my experience, so it would be nice to get a better grasp on it.

errordeveloper commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -65,22 +65,27 @@ func NewCmdInit(out io.Writer, s *kubeadmapi.KubeadmConfig) *cobra.Command {
+			kubeadmapi.DefaultServicesCIDRString,
+	)
+	cmd.PersistentFlags().IPNetVar(
+		&s.InitFlags.PodNetwork.CIDR, "cluster-cidr", net.IPNet{},
```
I think the flag should be `--pod-cidr`, because `--cluster-ip-range` in `kube-apiserver` means what we have as `--service-cidr`. I think pods and service are a better grouping, and the exact flags `kube-apiserver` takes aren't as clear, but we shouldn't conflate things even further. We are likely going to revisit the flags throughout alpha phase, so it's not set in stone yet, but please change this to `--pod-cidr` (or `--pod-network-cidr`).

errordeveloper commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -238,13 +237,26 @@ func getComponentCommand(component string, s *kubeadmapi.KubeadmConfig) (command
+
+			// Only append the --cloud-config option if there's a such file
+			if _, err := os.Stat(DefaultCloudConfigPath); err == nil {
+				command = append(command, "--cloud-config="+DefaultCloudConfigPath)
```
Looks like you have missed another instance of this down below...

errordeveloper reviewed (changes requested):
Please see fee comments regarding flag and the description, and one little artefact of refactoring `getComponentCommand`...

errordeveloper commented:
As we would like to progress towards getting code into master, I'm going to merge this and address  my comments above myself.

taimir commented:
I just took action on your nits, do you want me to quickly open another PR?

taimir commented:
Feel free to fix them yourself, I definitely don't want to slow you down.
